tests/driver_soft_uart/main.c

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ * @brief       Manual test application for UART peripheral drivers
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "board.h"
+#include "shell.h"
+#include "thread.h"
+#include "msg.h"
+#include "ringbuffer.h"
+#include "soft_uart.h"
+#include "soft_uart_params.h"
+
+#ifndef SHELL_BUFSIZE
+#define SHELL_BUFSIZE       (128U)
+#endif
+#ifndef UART_BUFSIZE
+#define UART_BUFSIZE        (128U)
+#endif
+
+#define PRINTER_PRIO        (THREAD_PRIORITY_MAIN - 1)
+#define PRINTER_TYPE        (0xabcd)
+
+typedef struct {
+    char rx_mem[UART_BUFSIZE];
+    ringbuffer_t rx_buf;
+} uart_ctx_t;
+
+static uart_ctx_t ctx[SOFT_UART_NUMOF];
+
+static kernel_pid_t printer_pid;
+static char printer_stack[THREAD_STACKSIZE_MAIN];
+
+#ifdef MODULE_SOFT_UART_MODECFG
+static uart_data_bits_t data_bits_lut[] = { UART_DATA_BITS_5, UART_DATA_BITS_6,
+                                            UART_DATA_BITS_7, UART_DATA_BITS_8 };
+static int data_bits_lut_len = ARRAY_SIZE(data_bits_lut);
+
+static uart_stop_bits_t stop_bits_lut[] = { UART_STOP_BITS_1, UART_STOP_BITS_2 };
+static int stop_bits_lut_len = ARRAY_SIZE(stop_bits_lut);
+#endif
+
+static int parse_dev(char *arg)
+{
+    unsigned dev = atoi(arg);
+    if (dev >= SOFT_UART_NUMOF) {
+        printf("Error: Invalid UART_DEV device specified (%u).\n", dev);
+        return -1;
+    }
+
+    return dev;
+}
+
+static void rx_cb(void *arg, uint8_t data)
+{
+    uart_t dev = (uart_t)arg;
+
+    ringbuffer_add_one(&(ctx[dev].rx_buf), data);
+    if (data == '\n' || ringbuffer_full(&(ctx[dev].rx_buf))) {
+        msg_t msg;
+        msg.content.value = (uint32_t)dev;
+        msg_send(&msg, printer_pid);
+    }
+}
+
+static void *printer(void *arg)
+{
+    (void)arg;
+    msg_t msg;
+    msg_t msg_queue[8];
+    msg_init_queue(msg_queue, 8);
+
+    while (1) {
+        msg_receive(&msg);
+        uart_t dev = (uart_t)msg.content.value;
+        int c;
+
+        printf("Success: UART_DEV(%i) RX: \n[", dev);
+        do {
+            c = ringbuffer_get_one(&(ctx[dev].rx_buf));
+            if (c == '\n') {
+                puts("]\\n");
+            }
+            else if (c >= ' ' && c <= '~') {
+                printf("%c", c);
+            }
+            else {
+                printf("0x%02x", (unsigned char)c);
+            }
+        } while (c > 0 && c != '\n');
```

Comment:
```suggestion
        } while (c >= 0 && c != '\n');
```

The ringbuffer will return `-1` when `ringbuffer_get_one()` is called while the buffer is empty. Sending some zero bytes breaks the test application.

Reply:
Sending a bunch of `'\n'` would still do so. I think it should just be `while (c >= 0);`

Reply:
No, a bunch of `'\n'` would result in a message being send to the printer for each char. That should not deadlock the test.